### PR TITLE
ref(py3): pre-commit upgrade + sweeping rerun + __future__ removals

### PR DESCRIPTION
**Pull request by joshuarli**
For anyone rebasing/merging against this when it lands, you'll need to `make setup-git` and run pre-commit on the files you changed in your branch. Something like `pre-commit run --files $(git whatchanged --name-only --pretty="" master... | sort | uniq)`.

This upgrades pre-commit and its hooks, moves it into a `requirements-pre-commit.txt` for git hook notifications on updates, and reruns pre-commit on everything. Most notably, black's target-version py36 removes `u` prefixes. flake8 is now run on `bin/` and I manually made those fixes. Also removes all `__future__`s (in our case we can remove everything).

### CODE REVIEW

**Comment by joshuarli** [comment]
Sorry for the notifications, codeowners. lol

image

**Review by untitaker — APPROVED** [review]
unsubscribe

File: bin/dump-command-help [hunk 1]
```diff
@@ -1,4 +1,6 @@
 #!/usr/bin/env python
+from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (joshuarli) [reply bin/dump-command-help:2]: I think so (don't know what it means right now), might as well get that in this PR while we have this big diff.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -12,24 +12,22 @@ exclude: >
...
+-   repo: local
     hooks:
+    # Configuration for black exists in pyproject.toml.
     - id: black
+      name: black
+      entry: black
+      language: python
       types: [python]
-      exclude: (migrations/)
--   repo: local
```
Comment (armenzg) [inline .pre-commit-config.yaml:21]: I'm trying to follow along. Did you mainly upgrade packages (and adjusted flags) and move the `local` repo at the top?

**Comment by billyvg** [comment]
Do we need to do this for getsentry as well?

File: Makefile [hunk 3]
```diff
@@ -79,8 +79,7 @@ setup-git: ensure-venv setup-git-config
 	@echo "--> Installing git hooks"
 	mkdir -p .git/hooks && cd .git/hooks && ln -sf ../../config/hooks/* ./
 	@python3 -c '' || (echo 'Please run `make setup-pyenv` to install the required Python 3 version.'; exit 1)
-	@# pre-commit loosely pins virtualenv, which has caused problems in the past.
-	$(PIP) install "pre-commit==1.18.2" "virtualenv==20.0.32"
+	$(PIP) install -r requirements-pre-commit.txt
```
Comment (armenzg) [inline Makefile:82]: Nice!

**Comment by joshuarli** [comment]
> Do we need to do this for getsentry as well?

Yes, I'm on it.

File: bin/dump-command-help (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshuarli) [reply bin/dump-command-help:2]: Oh haha, looks like this became default py3 behavior. Yeah, I'll get this in here.

File: config/hooks/post-merge [hunk 4]
```diff
@@ -9,11 +9,12 @@ trap "rm -f ${files_changed_upstream}" EXIT
 
 git diff-tree -r --name-only --no-commit-id ORIG_HEAD HEAD > "$files_changed_upstream"
 
-grep -E --quiet 'requirements.*\.txt' "$files_changed_upstream" && py="install-py-dev "
-grep -E --quiet 'yarn.lock' "$files_changed_upstream"           && js="install-js-dev "
-grep -E --quiet 'migrations' "$files_changed_upstream"          && migrations="apply-migrations "
+grep -E --quiet 'requirements-pre-commit\.txt' "$files_changed_upstream" && pc="setup-git "
```
Comment (armenzg) [inline config/hooks/post-merge:12]: Beauty.

**Review by armenzg — APPROVED** [review]
I did not look at anything under `src/` or `tests/`.
Thanks for handling this!

File: bin/dump-command-help (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshuarli) [reply bin/dump-command-help:2]: `fastmod 'from __future__ import absolute_import\n' ''` :)

File: .pre-commit-config.yaml (hunk 2 above)
Comment (joshuarli) [reply .pre-commit-config.yaml:21]: The configuration is consolidated in `pyproject.toml` for black and `setup.cfg` for flake8. And yes to everything.

File: config/hooks/post-merge (hunk 4 above)
Comment (joshuarli) [reply config/hooks/post-merge:12]: That's right Armen.

**Comment by joshuarli** [comment]
Oh, so since the file filtering is done by pre-commit I'll have to move those ignore patterns over from black configuration. Migration files were changed: `git whatchanged --name-only --pretty="" master... | sort | uniq | grep src/sentry/migrations`

@getsentry/owners-migrations do we want black to touch migrations? I'm assuming not.

**Comment by wedamija** [comment]
> 
> @getsentry/owners-migrations do we want black to touch migrations? I'm assuming not.
> 

It's fine either way. It won't cause any problems if you reformat them all, they won't re-run or anything like that.

**Comment by joshuarli** [comment]
> > @getsentry/owners-migrations do we want black to touch migrations? I'm assuming not.
> 
> It's fine either way. It won't cause any problems if you reformat them all, they won't re-run or anything like that.

Oh cool, I'll just keep formatting for them on then, since sometimes you need to manually author parts of migrations so black'd be helpful there too.

**Review by wedamija — APPROVED** [review]
I hand reviewed all 2780 files, lgtm

(this is a lie)

**Comment by joshuarli** [comment]
I'm probably going to wait for https://github.com/getsentry/sentry/pull/23195. @wedamija would also like your review on https://github.com/getsentry/getsentry/pull/4957.

**Comment by joshuarli** [comment]
Oh, might as well just remove all the `__future__`s. All the ones we have are not necessary in Python 3.6, you can refer to Python's `Lib/__future__.py`.

**Comment by joshuarli** [comment]
```
git fetch origin
git merge -Xtheirs origin/master
...let pre-commit run on the merge commit.
```

**Comment by joshuarli** [comment]
Ah backend lint cancelled at 3 min cap... @billyvg FYI this did not result in a "failure", I had to notice this.

image

Bumping to 10 for now.

**Comment by billyvg** [comment]
> Ah backend lint cancelled at 3 min cap... @billyvg FYI this did not result in a "failure", I had to notice this.
> 
> 
> 
> image
> 
> 
> 
> Bumping to 10 for now.

Yeah cancelled jobs have a subtle gray icon but will count as a failure if the check is required (which can't happen on gs)

**Comment by joshuarli** [comment]
lol

```
examples/oauth2_consumer_webserver/app.py:1:1: B317: Use ``from sentry.utils import json`` instead.
examples/oauth2_consumer_webserver/app.py:1:1: B317: Use ``from sentry.utils import json`` instead.
```

but other than this we lookin' good

**Comment by joshuarli** [comment]
Uhh, lol, guess I went at it too hard today.

```
Fetching list of changed files for PR#23197 from Github API
Error: API rate limit exceeded for installation ID 1204373.
```

**Pull request merged**